fem/lininteg_domain_grad.cpp

```diff
@@ -0,0 +1,424 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "fem.hpp"
+#include "../fem/kernels.hpp"
+#include "../general/forall.hpp"
+
+namespace mfem
+{
+
+template<int T_D1D = 0, int T_Q1D = 0> static
+void DLFGradAssemble2D(const int vdim, const int ne, const int d, const int q,
+                       const int *markers, const double *b, const double *g,
+                       const double *jacobians,
+                       const double *weights, const Vector &coeff, double *y)
+{
+   const auto F = coeff.Read();
+   const auto M = Reshape(markers, ne);
+   const auto B = Reshape(b, q, d);
+   const auto G = Reshape(g, q, d);
+   const auto J = Reshape(jacobians, q, q, 2,2, ne);
+   const auto W = Reshape(weights, q, q);
+   const bool cst = coeff.Size() == vdim*2;
+   const auto C = cst ? Reshape(F,2,vdim,1,1,1) : Reshape(F,2,vdim,q,q,ne);
+   auto Y = Reshape(y, d,d, vdim, ne);
+
+   MFEM_FORALL_2D(e, ne, q, q, 1,
+   {
+      if (M(e) == 0) { return; } // ignore
+
+      constexpr int Q = T_Q1D ? T_Q1D : MAX_Q1D;
+      constexpr int D = T_D1D ? T_D1D : MAX_D1D;
+
+      MFEM_SHARED double sBGt[2][Q*D];
+      MFEM_SHARED double sQQ[2][Q*Q];
+      MFEM_SHARED double sDQ[2][D*Q];
+
+      const DeviceMatrix Bt(sBGt[0], q, d);
+      const DeviceMatrix Gt(sBGt[1], q, d);
+      kernels::internal::LoadBGt<D,Q>(d, q, B, G, sBGt);
+
+      const DeviceMatrix QQ0(sQQ[0], q, q);
+      const DeviceMatrix QQ1(sQQ[1], q, q);
+
+      const DeviceMatrix DQ0(sDQ[0], d, q);
+      const DeviceMatrix DQ1(sDQ[1], d, q);
+
+      for (int c = 0; c < vdim; ++c)
+      {
+         const double cst_val0 = C(0,c,0,0,0);
+         const double cst_val1 = C(1,c,0,0,0);
+
+         MFEM_FOREACH_THREAD(x,x,q)
+         {
+            MFEM_FOREACH_THREAD(y,y,q)
+            {
+               const double w = W(x,y);
+               const double J11 = J(x,y,0,0,e);
+               const double J21 = J(x,y,1,0,e);
+               const double J12 = J(x,y,0,1,e);
+               const double J22 = J(x,y,1,1,e);
+               const double u = cst ? cst_val0 : C(0,c,x,y,e);
+               const double v = cst ? cst_val1 : C(1,c,x,y,e);
+               QQ0(y,x) = w * (J22*u - J12*v);
```

Comment:
This may help a bit:
```suggestion
               // QQ = w * det(J) * J^{-1} . C = adj(J) . { u, v }
               QQ(y,x) = w * (J22*u - J12*v);
```

Reply:
[done](https://github.com/mfem/mfem/pull/2568/commits/71e00731066a60e787f964b784efa699ce3bf509)